Bump astral to 2.1

alucryd opened the pull request:
## Proposed change

Bump astral to 2.1

## Type of change

- [x] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

Astral 2.x introduces a new LocationInfo class, uses dataclasses and refactored a lot of code. Tried to use the new functions to the same extent as the old one to the best of my knowledge.

Lots of tests are failing on the dev HEAD, using `tox -e py38`, even without these changes so I'm not sure whether I can check that local tests pass. FWIW the sun component worked fine in my local testing.

## Checklist

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

alucryd committed:
bump astral to 2.1

homeassistant commented:
Hello @alucryd,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`1f144d2439357cd3fccf80a3fa306e856f78d38d`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;32270&#x2F;commits&#x2F;1f144d2439357cd3fccf80a3fa306e856f78d38d) - This commit has something that looks like an email address (maxime.gauduin@majorel.biz). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

homeassistant commented:
Hi @alucryd, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#32270).

Once you do that we will be able to review and accept this pull request.

Thanks!

balloob commented on homeassistant/components/sun/__init__.py:
```diff
@@ -164,7 +165,7 @@ def update_events(self, utc_point_in_time):
         self.location.solar_depression = -10
         self._check_event(utc_point_in_time, "dawn", PHASE_SMALL_DAY)
-        self.next_noon = self._check_event(utc_point_in_time, "solar_noon", None)
+        self.next_noon = self._check_event(utc_point_in_time, "noon", None)
```
are `solar_noon` and `noon` equivalent?

Swamp-Ig replied:
I don't think so. Solar noon is when the sun is highest in the sky. Noon is 12:00.

alucryd replied:
The docstring for the noon function indicates: `Calculates the solar noon (the time when the sun is at its highest point.)` and reverse for midnight.

MartinHjelmare replied:
`noon` and `midnight` were renamed here:
https://github.com/sffjunkie/astral/commit/7e061884b72d3e4c77c34c665607db2be3ed4612

alucryd commented:
Guys, would you mind having a look at this PR? Would hate for it to be closed automatically...

MartinHjelmare commented:
Please rebase on latest dev branch. If tests pass I think we can merge.

MartinHjelmare commented:
I think you need to rebase from lastest dev branch, and remove any non related commits. The branch is now not clean.

alucryd commented:
Yeah, not sure why rebase messed up like that. In any case there are some test failures that are caused by these changes so I'll need to go over them. Will clean my branch while I'm at it.

alucryd commented on tests/components/sun/test_init.py:
```diff
@@ -151,10 +137,10 @@
     assert dt_util.parse_datetime(
         state.attributes[sun.STATE_ATTR_NEXT_RISING]
-    ) == datetime(2016, 7, 25, 23, 23, 39, tzinfo=dt_util.UTC)
+    ) == datetime(2016, 7, 24, 22, 59, 45, 689645, tzinfo=timezone.utc)
```
astral 2.1 returns these numbers now, double checked by using the library in a standalone test script, hope I'm not doing it wrong. Also the date parser returns datetime.timezone.utc, not pytz.utc, looks like ciso8601 does that.

MartinHjelmare replied:
Unless the input to the test has changed the output shouldn't change. I think something is wrong.

alucryd replied:
I agree, but you may verify these values using this very simple script: https://paste.xinu.at/rzpYfNPDbf3R2B9r/

Either the new astral is wrong and the tests should stay as is, or the old astral was wrong and the tests need to be adapted.

alucryd replied:
It was up on my face the whole time and I didn't notice, but the sun rises before it sets, even though it is already up, so there's definitely sth wrong with the new astral.

Found an issue about it: https://github.com/sffjunkie/astral/issues/34

Swamp-Ig replied:
The same issue has been there for a while, I had a look at fixing it ages ago but I didn't understand the maths involved. It's only in some polar regions that this happens.

alucryd commented:
Cleaned everything up and started looking at the tests. There were differences in the numbers returned in the norway test for the sun component, used those numbers instead since this is what the library is returning now if I follow their documentation's example. Would be great if another pair of eyes could confirm this.

Not sure why the automation tests for the sun fail, but it could the same root cause as the above. If that's the case and the library does return different numbers, I'll have to tweak the hardcoded datetimes there as well.

alucryd commented:
Okay I'm down to one failed test on the tod component with that famous norwegian city. Would be great if the component owner could have a look, I don't quite understand some of the logic of the component, and this particular test itself.

A few thoughts, the first assert seems wrong to me, at the specified date the sun is up, so it is after sunrise and before sunset, yet the assert checks `state == 'off'`. There's also what seems to be an unfortunate approximation in the code, if after and before event dates are after the current date, the code arbitrarily substracts one day to both dates, instead of getting the real dates which happen long before just one day prior. I'm not familiar with this so the owner will surely know better.

bdraco committed:
Update cast to use shared zeroconf (#35570)

* Update cast to use the shared zeroconf instance

* Add zeroconf to after_dependencies

* Bump version to 5.2.0

luizribeiro commented:
What is blocking us from moving forward here? Would be great to upgrade this library so we're not relying on such an old version

fredrike committed:
Fix Daikin duplicate entries (#35833)

Emilv2 committed:
Fix Delijn sensor naming (#35789)

uvjustin committed:
Handle None received from pyforked-daapd (#35830)

* Handle None received from API in forked-daapd

* Bump pyforked-daapd version in requirements

* Add test

bachya committed:
Fix Prezzibenzina doing I/O in the event loop (#35881)

* Fix Prezzibenzina doing I/O in the event loop

* Linting

hunterjm committed:
fix mjpeg issue along with some cameras not returning event capabilities properly (#35885)

bdraco committed:
Ensure http can startup if homekit fails to load (#35888)

* Ensure HomeAssistant can startup if homekit fails to load

* Update homeassistant/components/logbook/manifest.json

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

bdraco committed:
Fix legacy Hunter Douglas PowerView devices (#35895)

These devices are missing firmware information
as the 1.0 firmware did not provide it.

ctalkington committed:
Update rokuecp to 0.4.1 (#35899)

* update rokuecp to 0.4.1

* Update requirements_all.txt

* Update requirements_test_all.txt

bramkragten committed:
Updated frontend to 20200519.1 (#35877)

bdraco committed:
Ensure storage write consume the data under the lock (#35889)

If two writes trigger at the same time the data would already
be consumed.

bdraco committed:
Homekit should skip devices that are missing in device registry (#35857)

* Homekit should skip devices that are missing in device registry

* Add test for this failure state

danielperna84 committed:
Fix light profiles for HomeMatic lights (#35882)

pvizeli committed:
Bump version 0.110.1

bachya committed:
Auto-level AirVisual API calls (#34903)

pvizeli committed:
Merge pull request #35907 from home-assistant/rc

0.110.1

bramkragten committed:
Update frontend to 20200519.3 (#35925)

* Updated frontend to 20200519.2

* Updated frontend to 20200519.3

fredrike committed:
Fix Daikin AC integration for AirBase units (#35952)

Commit:
Fix typo in conf[CONF_SSL] (#35946)

hunterjm committed:
Bump pyAV and close unclosed outputs (#35960)

* bump pyAV and close unclosed outputs

* skip stream from coverage for now

* fix divide by zero error

bramkragten committed:
Update frontend to 20200519.4 (#35987)

MartinHjelmare committed:
Bump python-openzwave-mqtt to 1.0.2 (#35980)

MartinHjelmare committed:
Fix lutron_caseta setup options (#35974)

hunterjm committed:
Fix ONVIF Transport (#35932)

* allow lib to create AsyncTransport

* fix transport close issue

bdraco committed:
Ensure homekit functions if numpy is unavailable (#35931)

hunterjm committed:
Fix ONVIF PTZ and profile encoding issues (#36006)

* allow lib to create AsyncTransport

* fix transport close issue

* fix zoom only cameras without PTZ presets

* catch profiles without encoding configuration

* also catch ServerDisconnectedError for ptz

flz committed:
Fix iaqualink sensors (#36000)

* iaqualink: small sensor fixes

* Re-add device_class, fix type hints.

flz committed:
Bump iaqualink to 0.3.3 (#35999)

alengwenus committed:
Fix device_registry cleanup behavior (#35977)

* Fix: Only decives which are not referenced by an entity or a config_entry are removed

* Adapted test for async_cleanup

* Changed variable names

ctalkington committed:
Fix roku play/pause support (#35991)

felipediel committed:
Handle StorageError in the Broadlink integration (#35986)

soldag committed:
Migrate rpi_gpio_pwm to extend LightEntity instead of Light (#36028)

bdraco committed:
Use a single service browser for zeroconf discovery (#35997)

chmielowiec committed:
Fix service registration supported features check (#35718)

bdraco committed:
Fix shade compatibility with hunter douglas powerview 1.0 hubs (#36040)

frenck committed:
Upgrade hass-nabucasa to 0.34.3 (#36025)

Jc2k committed:
Bump aiohomekit (#36041)

ctalkington committed:
Improve ipp unique id parsing (#35959)

fredrike committed:
Bump tellduslive version (#36048)

frenck committed:
Bumped version to 0.110.2

Danielhiversen committed:
Fix opengarage async_setup_platform (#36075)

* opengarage async_setup_platform

* async_add_entities

frenck committed:
Merge pull request #36070 from home-assistant/rc

hunterjm committed:
Fix ONVIF config entry unique ID (#36008)

* fallback to device serial number if no mac available

* make password optional to fix #35904

* update tests to reflect new flow

* fix snake case and AsyncMock

* add comments around why weird things are being done

Julius2342 committed:
Bump pyvlx to 0.2.16 (#35971)

hunterjm committed:
guard against missing topic (#36108)

hunterjm committed:
fix preset warning (#36110)

ctalkington committed:
Update rokuecp to 0.4.2 (#36102)

* update rokuecp to 0.4.2

* Update requirements_all.txt

* Update requirements_test_all.txt

ctalkington committed:
Fix roku play/pause during standby (#36096)

Minims committed:
Fix onvif snapshot for Sricam SP009 (#36095)

isk0001y committed:
Re-read last imap_email_content email when no change (#36065)

bdraco committed:
Remove unsupported stop feature with Hunter Douglas Powerview 1.0 Hubs (#36129)

bdraco committed:
Ensure homekit bridge state is restored before creating devices (#36098)

* Ensure homekit bridge state is restored before creating devices

* Tests to ensure homekit device registry entry is stable

* remove stray continue

balloob committed:
Fix client ID lookup for official apps (#36131)

Tho85 committed:
Fix emulated_hue compatibility with older devices (#36090)

* Fix emulated_hue compatibility with older devices

* Fix test ugliness

* Fix pylint errors

bachya committed:
Fix bugs with AirVisual auto-leveling API (#36097)

* Fix bugs with AirVisual auto-leveling API

* Code review

* Code review

bramkragten committed:
Update frontend to 20200519.5 (#36154)

frenck committed:
Bumped version to 0.110.3

frenck committed:
Merge pull request #36155 from home-assistant/rc

Commit:
Fix cloud connection within API (#36158)

gadgetchnnel committed:
Check todoist due date is not None in async_get_events (#36140)

* Check that due date is not None

Check that due date is not None, prevents taks without due dates from breaking Calendar API

* Invert None check to reduce indentation

emontnemery committed:
Let PAHO MQTT client handle connection to MQTT server (#35983)

* Let PAHO client handle connection to MQTT server

bdraco committed:
Add ability to ignore heos discovery (#34653)

* Add ability to ignore heos discovery

* Fetch player_id, update tests

* Handle failure state

* Update tests as there are two players in the mock now

* Adjust and add more tests

* Strip out player id lookup

* reverts per review

* one more revert

JeffLIrion committed:
Fix Android TV icon when screencap option is disabled (#35710)

* Don't return a media image hash if the screencap config option is False

* 1-liner

hunterjm committed:
Fix empty preset element in ONVIF response (#36182)

balloob committed:
Revert DSMR not calling entity methods (#36179)

emontnemery committed:
Bump zeroconf, pychromecast. Log if zeroconf.get_service_info fails. (#36185)

ctalkington committed:
Fix roku select source with app ids (#36191)

bachya committed:
Prevent AirVisual from polling (#36199)

* Prevent AirVisual from polling

* Docstring

emontnemery committed:
Fix custom position range (#36222)

emontnemery committed:
Correct MQTT device trigger reconfiguration with same topic (#36234)

balloob committed:
Bumped version to 0.110.4

balloob committed:
Merge pull request #36238 from home-assistant/rc

alucryd committed:
bump astral to 2.2

bdraco commented:
This PR needs to be rebased as there are unrelated changes.

https://developers.home-assistant.io/docs/development_catching_up/

alucryd commented:
There's no point in wasting time on this until upstream astral has been fixed in places like Norway.

MartinHjelmare commented:
I think we can close here until the library is fixed.

@alucryd if you close, you can reopen yourself and we keep the PR thread history intact.

alucryd commented:
@MartinHjelmare Sure, let's do that.

amitfin commented:
@alucryd, IIUC, this PR seems to be blocked because of https://github.com/sffjunkie/astral/issues/34. However, that issue was opened on Jun 5, 2019, while astral version 2.0 was [released on Feb 10, 2020](https://github.com/sffjunkie/astral/releases). So, this issue is not a regression of a newer version of the library, as the issue was opened against astral version 1.10.1, which is the version currently being used by HA. Furthermore, I believe that https://github.com/sffjunkie/astral/issues/34 was actually [fixed in astral version 2.0](https://github.com/sffjunkie/astral/issues/34#issuecomment-720915926), which is another good reason to revive and push forward this PR.

(Context: I'm trying to push for upgrading astral version in HA because we have a dependency on the newer astral lib with HA's "jewish_calendar" integration and its "hdate" underline library.)

alucryd commented:
I'll try to get back to this during the week, but last time I tried astral 2.1 it broke every single unit test involving hardcoded sun positions and it wasn't just in arctic regions. Will see if 2.2 fares better.

amitfin commented:
@alucryd, did you give it a try? Thanks!
